### ISSUE TITLE
add Takagi decomposition routine 

### ISSUE THREAD

**Issue by mgbukov** [scipy/scipy#16844]
### Is your feature request related to a problem? Please describe.

It'd be nice to add to `scipy.linalg` a function that computes the [Takagi decomposition](https://en.wikipedia.org/wiki/Matrix_decomposition#Takagi's_factorization) (a.k.a. symmetric SVD). In short, for a complex-valued square matrix $M$ which is symmetric (but not necessarily hermitian): $M = M^T$, the Takagi decomposition states that there exists a unitary matrix $V$, such that $M = VDV^T$, where $D$ is a diagonal matrix with nonnegative entries. 

### Describe the solution you'd like.

Ideally, scipy would have an efficient version for the algorithm instead of the code suggested below, see e.g., [IAENG International Journal of Applied Mathematics, 48:3, IJAM_48_3_08](http://www.iaeng.org/IJAM/issues_v48/issue_3/IJAM_48_3_08.pdf)

### Describe alternatives you've considered.

Here's a slight modification of an (admittedly inefficient) way to do the decomposition that I found on the web:

```
def takagi(A):
    """Extremely simple and inefficient Takagi factorization of a 
    symmetric, complex matrix A.  Here we take this to mean A = U D U^T
    where D is a real, diagonal matrix and U is a unitary matrix.  There
    is no guarantee that it will always work. """
    # Construct a Hermitian matrix.
    H = A.T.conj() @ A
    # Calculate the eigenvalue decomposition of the Hermitian matrix.
    # The diagonal matrix in the Takagi factorization is the square
    # root of the eigenvalues of this matrix.
    u, lam = scipy.linalg.eigh(H)
    # The "almost" Takagi factorization.  There is a conjugate here
    # so the final form is as given in the doc string.
    T = np.einsum('ji,jk,ki->i', u,A,u ) 
    # T is diagonal but not real.  That is easy to fix by a
    # simple transformation which removes the complex phases
    # from the resulting diagonal matrix.
    c = np.exp(0.5j*jnp.angle(T))
    U = np.einsum('ij,j->ij',u.conj(),c)
    # Now A = np.dot(U, np.dot(np.diag(np.sqrt(lam)),U.T))
    return np.sqrt(lam[::-1]), U[:,::-1]
```

which can be tested:

```
    np.set_printoptions(suppress=True,precision=4)
    np.random.seed(0)
    A = np.random.uniform(size=(4,4)) + 1j*np.random.uniform(size=(4,4))
    A += A.T

    lmbda,V = takagi(A)

    print(np.linalg.norm(A - V@np.diag(lmbda)@V.T))
```

**Comment by ev-br** [comment]
A drive-by comment: Please add a link to where this snippet comes from (and what is the license, if known).

**Comment by WarrenWeckesser** [comment]
It looks like it is from an answer on stackoverflow: https://stackoverflow.com/questions/19491936/complex-symmetric-matrices-in-python

**Comment by tupui** [comment]
> It looks like it is from an answer on stackoverflow: [stackoverflow.com/questions/19491936/complex-symmetric-matrices-in-python](https://stackoverflow.com/questions/19491936/complex-symmetric-matrices-in-python)

i.e. we cannot use it and need the approval from OP to relicense.

**Comment by mgbukov** [comment]
> It looks like it is from an answer on stackoverflow: https://stackoverflow.com/questions/19491936/complex-symmetric-matrices-in-python

that is the correct reference.

**Comment by ilayn** [comment]
Takagi decomposition is not guaranteed to exist for all symmetric complex arrays and also there are many issues to tackle to be production ready. 

I don't know any sources that offer signiificant performance boost compared to the regular SVD while still being reliable enough. Thus all in all, we don't have a strong use case for it to maintain a custom rolled-out version of it in my opinion.

**Comment by tupui** [comment]
Ok thank you all. Let's close this for now then.

**Comment by nquesada** [comment]
Hello everyone --- Would there still be interested in implementing this? In this reference we provide a good algorithm to do Autonne-Takagi decomposition of symmetric matrices (cf box 1): https://arxiv.org/abs/2403.04596 .
The algorithm is implemented in `thewalrus` [here](https://github.com/XanaduAI/thewalrus/blob/25b5f0820520865a311de2e7d44779e7338adc1d/thewalrus/decompositions.py#L149) .
The code is heavily tested there: https://github.com/XanaduAI/thewalrus/blob/master/thewalrus/tests/test_decompositions.py
Finally, in the arXiv preprint above we also give nice recipes for other more "exotic" matrix decompositions (Bloch-Messiah, Iwasawa, Williamson) involving symplectic matrices. Would there be any interest in having that in `scipy`? Just like Autonne-Takagi, these decompositions have been implemented, documented and tested in `thewalrus`.

**Comment by ilayn** [comment]
I don't think we can facilitate it here in my opinion due to reasons I gave above. It is too specialized and also it is not guaranteed to give an answer. It creates many tickets and frustration if it doesn't work in SciPy.

But it's just one vote so let's see what others think.

**Comment by nquesada** [comment]
Thanks for your prompt reply @ilayn . Can you point to a place in the literature where it is proven that "it is not guaranteed to give an answer"?

**Comment by ilayn** [comment]
Ikramov's paper if I remember correctly should include all the cases.

**Comment by ilayn** [comment]
The emphasis they have is the explicit construction of `W` in your notation. Here you leave it to success of the SVD convergence. In the literature, typically this decomposition algorithm is given only for invertible or unitary or some kind of other restrictions. I don't really remember what kind of counter-example I had in mind but I can't seem to figure out now either uniqueness or something like that. Though I remember having issues with outer Hankel factor problem or with the existence of square root but anyways. 

I wish I had some more time to look into it but maybe I am completely wrong anyways which is very often the case. So let me concede my argument so that it does not block the discussion here. 

However, the specialization part of my argument still holds in my opinion; if this algorithm always works then isn't it a one-liner `W = U @ sqrtm(U.T@V)` ? I can see that in real case, you are adjusting the ordering but still looks like a helper function to me.

**Comment by nquesada** [comment]
Thanks again for your prompt reply @ilayn . I am glad we are past the point of discussing the existence of this decomposition. You are correct that the implementation can be short and sweet. I think there is merit in having convenience short functions. My favourite example of a short sweet function that makes my life better is `fftfreq` in `numpy` (https://numpy.org/doc/stable/reference/generated/numpy.fft.fftfreq.html).
In any case the proposed `takagi` would be a little longer than the box I pasted above as for the case where the input matrix is real one can speed up the calculation a little bit.
